index.bs

```diff
@@ -147,35 +151,109 @@ Exactly how external changes are reflected in the data structures defined by thi
+
+  1. Let |childPath| be |child|'s [=FileSystemHandle/locator=]'s [=file system locator/path=].
+  1. Let |rootPath| be |root|'s [=FileSystemHandle/locator=]'s [=file system locator/path=].
+  1. If |childPath| is [=the same path as=] |rootPath|,
      [=/resolve=] |result| with an empty list, and abort.
-  1. Let |childPromises| be « ».
-  1. [=set/For each=] |entry| of |root|'s [=FileSystemHandle/entry=]'s [=children=]:
-    1. Let |p| be the result of [=file system entry/resolving=] |child| relative to |entry|.
-    1. [=list/Append=] |p| to |childPromises|.
-    1. [=Upon fulfillment=] of |p| with value |path|:
-      1. If |path| is not null:
-        1. [=list/Prepend=] |entry|'s [=file system entry/name=] to |path|.
-        1. [=/Resolve=] |result| with |path|.
-  1. [=Wait for all=] |childPromises|, with the these success steps:
-    1. If |result| hasn't been resolved yet, [=/resolve=] |result| with `null`.
+
+  1. If |rootPath|'s [=list/size=] is greater than |childPath|'s [=list/size=],
+     [=/resolve=] |result| with null, and abort.
+
+  1. [=list/For each=] |index| of |rootPath|'s [=list/indices=]:
+    1. If |rootPath|.\[[|index|]] is not |childPath|.\[[|index|]], then
+       [=/resolve=] |result| with null, and abort.
+
+  1. Let |relativePath| be an empty [=/list=].
+  1. [=list/For each=] |index| of [=the range=] from |rootPath|'s [=list/size=]
+     to |rootPath|'s [=list/size=], exclusive,
+     [=list/append=] |childPath|.\[[|index|]] to |relativePath|.
+
+  1. [=/Resolve=] |result| with |relativePath|.
+
 1. Return |result|.
 
 </div>
 
+A <dfn export>file system locator</dfn> represents a potential location of a
+[=/file system entry=]. A [=/file system locator=] is either a [=file locator=]
+or a [=directory locator=].
+
+Each [=/file system locator=] has an associated <dfn export for="file system locator" id=locator-path>path</dfn> (a [=/file system path=]),
+a <dfn export for="file system locator" id=locator-kind>kind</dfn> (a {{FileSystemHandleKind}}), and
+a <dfn export for="file system locator" id=locator-root>root</dfn> (an opaque string).
+
+A <dfn export>file locator</dfn> is a [=/file system locator=] whose
+[=file system locator/kind=] is {{FileSystemHandleKind/"file"}}.
+A <dfn export>directory locator</dfn> is a [=/file system locator=] whose
+[=file system locator/kind=] is {{FileSystemHandleKind/"directory"}}.
+
+A <dfn export>file system root</dfn> is an opaque string whose value is
+[=implementation-defined=].
+
+<p class=example id=example-locator>For a [=/file system locator=] |locator|
+whichs [=locate an entry|locates to=] a [=file entry=] |entry| that conceptually
+exists at the path `data/drafts/example.txt` relative to the root directory of
+the [=origin private file system=],
+|locator|'s [=file system locator/kind=] has to be {{FileSystemHandleKind/"file"}},
+|locator|'s [=file system locator/path=] might be « "`data`", "`drafts`", "`example.txt`" », and
```

Comment:
I say "might" here because it's still "implementation defined" how a "locator" maps to an "entry", and most of the methods (with `isSameEntry()` and `resolve()` being the exceptions) require the "entry" to exist to do anything useful. For example, take the `getFileHandle(name)` method. It's basically:

1. Given a "locator", locate `entry` (a dir entry)
2. Given `entry`, find the `child` whose name is `name`
3. Create a FileSystemFileHandle given `child`, which involves "getting the locator" of `child`

So nowhere in this flow do we ever specify appending `name` to `this`'s locator's path to get `child`'s locator's path

Relatedly, creating the root of an OPFS doesn't say anything about its locator's path

If we want to assert that the path above "has to be" `/data/drafts/example.txt` (in list form) then we could:
- ensure that handles acquired from various API surfaces are given a defined path
  - set the path of the root of an OPFS to a list containing one empty string
  - outside of the OPFS (e.g. for a file picked from a file picker), let the path be "implementation defined"
- create handles from existing handles by manipulating the existing handle's locator's path
  - add new versions of the "create a new FileSystem{File|Directory}Handle" algorithms that append to the path of the existing handle

I'm happy to make this change (either here or in a follow-up), but this is where `getFileHandle()` and `getDirectoryHandle()` get thorny, as @szewai pointed out https://github.com/whatwg/fs/issues/59#issuecomment-1402897170). Arguably, these methods could ignore the state on disk and just always return a handle whose locator has had its path appended to, without bothering to check what exists on disk....

... but this would be a breaking change, and a pretty substantial one where `{ create: true }` is currently passed

For `{ create: false }` with `dir` that exists at path "/testing":
```
await dir.remove();
// Before: throws a NotFoundError
// After: succeeds - handle's path is "/testing/example.txt"
let handle = await dir.getFileHandle('example.txt', { create: false });
```
This is a web-exposed change, but arguably not that bad, especially if we add something like #80. Something that threw an exception before would longer throw.

For `{ create: true }` with `dir` that exists at path "/testing" if we deprecate the "create" option:
```
// Before: creates a new file at "/testing/example.txt"
// After: handle's path is "/testing/example.txt" - but no file is created
let handle = await dir.getFileHandle('example.txt', { create: true });

// At this point, any method which cares about what's on disk now fails

// Before: creates a SyncAccessHandle
// After: throws a NotFoundError
let accessHandle = await handle.createSyncAccessHandle();

// Before: gets the file's contents
// After: throws a NotFoundError
let file = await handle.getFile();
```

I worry that deprecating this option would break _a lot_ of existing code. Something that [works now](https://webkit.org/blog/12257/the-file-system-access-api-with-origin-private-file-system/#create:~:text=Creating%20a%20directory%20or%20a%20file) would consistently throw

My tentative thoughts are that we should just leave this as "might" in this PR, then make the changes above to turn this into a "has to be", then look into making `getFileHandle()` and `getDirectoryHandle()` (perhaps by adding an additional flag? We can hash that out later :) )

Thoughts?

Reply:
Let me take a step back here to ensure I understand.

* `getFileHandle()` is exposed on `FileSystemDirectoryHandle`.
* A `FileSystemDirectoryHandle` is represented by a locator and as such when a method is invoked upon it that might fail if the locator no longer has a corresponding entry.

Why could we not append the given name to the locator's path and then see if an entry exists there? I think we can keep the existing behavior here, even if it's not necessarily the most natural given the "new" underpinnings.

---

I think for the OPFS root the path would indeed have to be the empty list. I'm not sure about making it contain an empty string. That only seems sensible if we did directories that way and you decided we wouldn't do that.

---

For file pickers you'd want the root to be a unique identifier and the path to be a list containing the file name, I think. At least by default `<input type=file>` shouldn't start exposing more about the user's machine.

Reply:
> Let me take a step back here to ensure I understand.
> 
> * `getFileHandle()` is exposed on `FileSystemDirectoryHandle`.
> * A `FileSystemDirectoryHandle` is represented by a locator and as such when a method is invoked upon it that might fail if the locator no longer has a corresponding entry.

Correct 👍 

> Why could we not append the given name to the locator's path and then see if an entry exists there? I think we can keep the existing behavior here, even if it's not necessarily the most natural given the "new" underpinnings.

That's basically what I was proposing on the other thread, so yeah that SGTM. I just wanted to explain what I understood of the other perspectives that were expressed (@szewai please correct me if I've misunderstood anything!)

The only nitpick here is that if `{ create: true }` currently fails rather than recursively create an entry. So we'd still need to check `this`'s locator first before appending the given name, or change the behavior

> I think for the OPFS root the path would indeed have to be the empty list. I'm not sure about making it contain an empty string. That only seems sensible if we did directories that way and you decided we wouldn't do that.

Yeah I agree that having a list with an empty string isn't great, but the empty string _is_ the name of the directory. We'd have to carve out an exception for OPFS roots in this (which isn't that big of a deal):

```
If [=locating an entry=] given [=/file system locator=] |locator| returns a
[=/file system entry=] |entry|, then |entry|'s [=file system entry/name=] must
be the same [=string=] as the last [=list/item=] of
|locator|'s [=file system locator/path=].
```

> For file pickers you'd want the root to be a unique identifier and the path to be a list containing the file name, I think. At least by default `<input type=file>` shouldn't start exposing more about the user's machine.

I agree that this would be ideal, but consider the following sequence:
- show a file picker and select `data/drafts/example.txt`
- show a directory picker and select `data`
- resolve `data/drafts/example.txt` relative to `data` (this should be `[ 'drafts', 'example.txt' ]`)

This won't resolve correctly if the "file system path" of the first file is just `[ 'example.txt' ]`

We could specify that the last item of the "file system path" should be the entry's name, but I think the rest of the path probably needs to be implementation-defined? And since the "file system path" isn't expected to be exposed to the end user, I don't think it should really matter?

Reply:
> The only nitpick here is that if { create: true } currently fails rather than recursively create an entry. So we'd still need to check this's locator first before appending the given name, or change the behavior

Can you elaborate on this, I'm not sure I follow.

> but consider the following sequence:

Well, you could resolve with null in this case. We don't have to tell the website more than the user told them. If you want to make it work you'd have to do away with the concept of a root I think (that would have to be part of the path), but I'm not necessarily convinced it needs to work.

Reply:
> Taking a step back... it seems like a bad idea to have different locators that point to the same entry anyways?

I'm not sure. I'll try to get some clarity on this.

> By the way, any thoughts on this?

It seems reasonable. We probably have to clarify somewhere that the name getter can return an invalid file name in this special case.

Reply:
> If we want to assert that the path above "has to be" `/data/drafts/example.txt` (in list form) then we could:
> * ensure that handles acquired from various API surfaces are given a defined path
>   * set the path of the root of an OPFS to a list containing one empty string
>   * outside of the OPFS (e.g. for a file picked from a file picker), let the path be "implementation defined"
> * create handles from existing handles by manipulating the existing handle's locator's path
>   * add new versions of the "create a new FileSystem{File|Directory}Handle" algorithms that append to the path of the existing handle

This is all included in the latest patch, aside from creating handles outside of the OPFS (which will have to be done in a follow-up on https://github.com/WICG/file-system-access, for which I _think_ I've exported all the terms I'll need... 🤞)

Also added a warning that the "file system path" shouldn't be exposed directly to the website process - it can only be implied by resolving the handle to a parent, as in this use case:

> * show a file picker and select `data/drafts/example.txt`
> * show a directory picker and select `data`
> * resolve `data/drafts/example.txt` relative to `data` (this should be `[ 'drafts', 'example.txt' ]`)